src/collider/BVHTree.cpp

```diff
@@ -322,91 +223,138 @@ uint32_t SingleBVHTree::Partition(SortKey *keys, uint32_t numKeys, const AABBd &
-		stack[stackLevel++] = node->kids[0];
+	// Simple O(n) sort algorithm to sort all objects according to side of pivot
+	uint32_t startIdx = 0;
+	uint32_t endIdx = numKeys;
+
+	// It is ~10% faster to sort the indices than to sort the whole AABB array
+	// (cache hit rate vs. memory bandwidth)
+	// Sorting in general is extremely fast.
+	while (startIdx < endIdx) {
+		if (keys[startIdx].center[axis] < pivot) {
+			startIdx++;
+		} else {
+			endIdx--;
+			std::swap(keys[startIdx], keys[endIdx]);
+		}
 	}
+
+	return startIdx;
 }
 
-double SingleBVHTree::CalculateSAH() const
+float BinnedAreaBVHTree::FindPivot(SortKey *keys, uint32_t numKeys, const AABBd &aabb, const AABBd *objAabbs, uint32_t axis, float &outCost) const
 {
-	double outSAH = 0.0;
-	auto *rootNode = GetNode(0);
-	std::vector<uint32_t> m_nodeStack = { 0 };
+	// Uses binned surface-area-heuristic construction:
+	// https://jacco.ompf2.com/2022/04/21/how-to-build-a-bvh-part-3-quick-builds/
 
-	while (!m_nodeStack.empty()) {
-		auto *node = GetNode(m_nodeStack.back());
-		m_nodeStack.pop_back();
+	SortBin bins[NUM_BINS] = {};
 
-		if (node->kids[0])
-			m_nodeStack.push_back(node->kids[0]);
-		if (node->kids[1])
-			m_nodeStack.push_back(node->kids[1]);
+	float boundsMin = ToSortSpace(aabb.min)[axis];
+	float boundsMax = ToSortSpace(aabb.max)[axis];
 
-		// surface area = 2 * width * depth * height
-		vector3d size = node->aabb.max - node->aabb.min;
-		double area = 2.0 * size.x * size.y * size.z;
+	float scale = NUM_BINS / (boundsMax - boundsMin);
+	float invScale = (boundsMax - boundsMin) / NUM_BINS;
 
-		// Cost function according to https://users.aalto.fi/~laines9/publications/aila2013hpg_paper.pdf Eq. 1
-		outSAH += (node->kids[0] ? 1.2 : 1.0) * area;
+	// Bin each object, using the sort key as its centroid
+	for (size_t i = 0; i < numKeys; i++) {
+		size_t bin_idx = std::min(NUM_BINS - 1, size_t((keys[i].center[axis] - boundsMin) * scale));
+		bins[bin_idx].objCount++;
+		bins[bin_idx].bounds.Update(objAabbs[keys[i].index]);
 	}
 
-	vector3d rootSize = rootNode->aabb.max - rootNode->aabb.min;
-	double rootArea = 2.0 * rootSize.x * rootSize.y * rootSize.z;
+	constexpr size_t MAX_BIN = NUM_BINS - 1;
 
-	// Perform 1 / Aroot * ( SAH sums )
-	outSAH /= rootArea;
+	float planeCost[MAX_BIN];
+	float bestCost = FLT_MAX;
+	float pivot = 0.0;
 
-	// Remove the (normalized) SAH cost of the root node from the result
-	// The SAH metric doesn't include the cost of the root node
-	outSAH -= 1.0;
+	AABBd leftBox = AABBd::Invalid();
+	AABBd rightBox = AABBd::Invalid();
+	uint32_t leftSum = 0;
+	uint32_t rightSum = 0;
 
-	return outSAH;
+	// Calculate the left-side SAH cost of each splitting plane
+	for (int i = 0; i < MAX_BIN; i++) {
+		leftSum += bins[i].objCount;
+		leftBox.Update(bins[i].bounds);
+		planeCost[i] = leftSum * leftBox.SurfaceArea();
+	}
+
+	// Calculate the right-side SAH cost of each splitting plane
+	for (int i = MAX_BIN - 1; i >= 0; i--) {
+		rightSum += bins[i].objCount;
```

Comment:
`bins[MAX_BIN]` not considered anywhere.

Reply:
Good catch, thank you! This was a refactoring/bugfix error late at night.